datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -1503,370 +1858,416 @@ pub async fn from_substrait_func_args(
 
 /// Convert Substrait AggregateFunction to DataFusion Expr
 pub async fn from_substrait_agg_func(
-    state: &dyn SubstraitPlanningState,
+    consumer: &impl SubstraitConsumer,
     f: &AggregateFunction,
     input_schema: &DFSchema,
-    extensions: &Extensions,
     filter: Option<Box<Expr>>,
     order_by: Option<Vec<SortExpr>>,
     distinct: bool,
 ) -> Result<Arc<Expr>> {
-    let args =
-        from_substrait_func_args(state, &f.arguments, input_schema, extensions).await?;
-
-    let Some(function_name) = extensions.functions.get(&f.function_reference) else {
+    let Some(fn_signature) = consumer
+        .get_extensions()
```

Comment:
You mentioned you'd prefer to not have the get_extensions(), I guess you could replace it with something like `get_function` which in DefaultSubstraitConsumer would just do this `extension.functions.get(..)`? Or did you have something else in mind?

Reply:
I was trying to figure out the right abstraction not just for resolving the function via the extension, but also mapping it into DataFusion. However I found the existing logic a little unwieldy to reason about. Specifically scalar functions first try and lookup the function by name, then look it up as an operator and then try and treat is a BuiltinExpr.

I want to think about function mapping in isolation, because I suspect it will be easier to figure out the right abstraction if we simplify the lookup logic first.

Reply:
> scalar functions first try and lookup the function by name, then look it up as an operator and then try and treat is a BuiltinExpr.

Yeah, I think this is a side-effect of DF not having (yet?) converged fully on UDFs. I think for aggregates and windows the built-in modes were already removed in favor of UDAF/UDWF but not for scalars.

But yep, totally good to do as a followup someday!